system/sqlstore.go

```diff
@@ -0,0 +1,194 @@
+package system
+
+import (
+	"bytes"
+	"database/sql"
+	"encoding/base32"
+	"encoding/gob"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/decred/dcrstakepool/models"
+	"github.com/go-gorp/gorp"
+	"github.com/gorilla/securecookie"
+	"github.com/gorilla/sessions"
+)
+
+// SQLStore stores gorilla sessions in a database.
+type SQLStore struct {
+	Options *sessions.Options
+	codecs  []securecookie.Codec
+	dbMap   *gorp.DbMap
+}
+
+// NewSQLStore returns a new SQLStore. The keyPairs are used in the same way as
+// the gorilla sessions CookieStore.
+func NewSQLStore(dbMap *gorp.DbMap, keyPairs ...[]byte) *SQLStore {
+	s := &SQLStore{
+		codecs: securecookie.CodecsFromPairs(keyPairs...),
+		dbMap:  dbMap,
+	}
+	// clean db of expired sessions once a day
+	go func() {
+		for {
+			time.Sleep(time.Hour * 24)
+			if err := s.destroyExpiredSessions(); err != nil {
+				log.Warn(err)
+			}
+		}
+	}()
+	return s
+}
+
+// Get returns a cached session.
+func (s *SQLStore) Get(r *http.Request, name string) (*sessions.Session, error) {
+	return sessions.GetRegistry(r).Get(s, name)
+}
+
+// New creates a new session for the given request r. If the request
+// contains a valid session ID for an existing, non-expired session,
+// then that session will be loaded from the database.
+func (s *SQLStore) New(r *http.Request, name string) (*sessions.Session, error) {
+	session := sessions.NewSession(s, name)
+	opts := *s.Options
+	session.Options = &opts
+	c, err := r.Cookie(name)
+	if err != nil {
+		if err == http.ErrNoCookie {
+			return session, nil
+		}
+		return session, err
+	}
+	err = securecookie.DecodeMulti(name, c.Value, &session.ID, s.codecs...)
+	if err != nil {
+		// these are not the sessions you are looking for
+		return session, nil
```

Comment:
I'm glad you spotted this.  I was debating what to do.  If there is a cookie with the same name "session" in the request it will error here. For example, if you were using a different branch and switched to this pr mid page, it will error because we were saving a map where we are now saving as string.  Then an error here is expected. But as there could be a real error here and it should be logged.
Is the Star Wars reference ok :P
It's doesn't return nil for the same reason as the other comment.

Reply:
I have changed this to log an error as info.  I'm not sure if that is optimal, let me think about it a little bit more.  Even if it did return an error, it would only be for logging purposes when we call New in the middleware.  In master all errors are being ignored anyway, so it's an improvement either way.

Reply:
After thinking a bit more, I think I like it like this. It's unlikely that an error here is something to worry about, but since I'm not sure what may happen in the real world I think it should be logged. So logging as info, while errors we are sure are problematic are returned and logged as warn later.